src/api/channel/by_gop.rs

```diff
@@ -34,50 +38,55 @@ impl<T: Pixel> SubGop<T> {
 }
 */
 
-// Extra
+// TODO: Make the detector logic fitting the model
 struct SceneChange {
-  frames: u64,
+  frames: usize,
   pyramid_size: usize,
-  min_key_frame_interval: u64,
-  max_key_frame_interval: u64,
+  processed: u64,
+  last_keyframe: u64,
+  detector: SceneChangeDetector,
 }
 
 impl SceneChange {
-  fn new(
-    pyramid_size: usize, min_key_frame_interval: u64,
-    max_key_frame_interval: u64,
-  ) -> Self {
-    Self {
-      frames: 0,
+  fn new(pyramid_size: usize, enc: &EncoderConfig) -> Self {
+    let seq = Arc::new(Sequence::new(enc));
+
+    let detector = SceneChangeDetector::new(
+      *enc,
+      CpuFeatureLevel::default(),
       pyramid_size,
-      min_key_frame_interval,
-      max_key_frame_interval,
-    }
+      seq,
+      true,
+    );
+
+    Self { frames: 0, pyramid_size, processed: 0, last_keyframe: 0, detector }
   }
 
   // Tell where to split the lookahead
-  // 7 is currently hardcoded, it should be a parameter
+  //
   fn split<T: Pixel>(
     &mut self, lookahead: &[Arc<Frame<T>>],
   ) -> Option<(usize, bool)> {
-    self.frames += 1;
+    self.processed += 1;
 
-    let new_gop = if self.frames < self.min_key_frame_interval {
-      false
-    } else if self.frames >= self.max_key_frame_interval {
-      self.frames = 0;
-      true
-    } else {
-      false
-    };
+    let new_gop = self.detector.analyze_next_frame(
+      &lookahead[self.frames..],
+      self.processed,
+      self.last_keyframe,
+    );
 
-    let len = lookahead.len();
+    if new_gop {
+      self.last_keyframe = self.processed;
+    }
 
-    if len > self.pyramid_size {
+    if self.frames > self.pyramid_size {
+      self.frames = 0;
       Some((self.pyramid_size, new_gop))
     } else if new_gop {
-      Some((len - 1, true))
+      self.frames = 0;
+      Some((self.frames, true))
```

Comment:
Note that `self.frames` is set to zero before being referenced in the return value. This is probably not correct.

Reply:
you found the bug :)